src/QirRuntime/lib/Tracer/README.md

```diff
@@ -0,0 +1,177 @@
+| `void __quantum__qis__on_operation_end(i64 %id)`      | Function to identify the end of a quantum module. The argument is a unique _id_ of the module and must match the _id_ supplied on start of the module. The call to the function will be inserted into QIR by the Q# compiler when Tracer is specified as the compilation target. |
+| `void __quantum__qis__single_qubit_op(i32 %id, i32 %duration, %Qubit* %q)` | Function for counting operations that involve a single qubit. The first argument is the id of the operation, as assigned by the client. Multiple intrinsics can be assigned the same id, in which case they will be counted together. The second argument is duration to be assigned to the particular invocation of the operation. |
+| `void __quantum__qis__multi_qubit_op(i32 %id, i32 %duration, %Array* %qs)` | Function for counting operations that involve multiple qubits.|
+| `void __quantum__qis__single_qubit_op__ctl(i32 %id, i32 %duration, %Array* %ctls, %Qubit* %q)` | Function for counting controlled operations with single target qubit. |
+| `void __quantum__qis__multi_qubit_op__ctl(i32 %id, i32 %duration, %Array* %ctls, %Array* %qs)` | Function for counting controlled operations with multiple target qubits. |
+| `%Result* @__quantum__qis__single_qubit_measure(i32 %id, i32 %duration, %Qubit* %q)` | Function for counting measurements of a single qubit. The user might assign different operation ids for different measurement bases. |
+| `%Result* @__quantum__qis__joint_measure(i32 %id, i32 %duration, %Array* %qs)` | Function for counting joint-measurements of qubits. The user might assign different operation ids for different measurement bases. |
+| `void __quantum__qis__swap(%Qubit* %q1, %Qubit* %q2)` | See [Special handling of SWAP](#special-handling-of-swap) for details. |
+
+_Note on operation ids_: The client is responsible for using operation ids in a consistent manner. Operations with the
+ same id will be counted by the tracer as the _same_ operation, even accross invocations with different number of target
+ qubits or when different functors are applied.
+
+## Native backing of the extension methods ##
+
+The Resource Tracer will reuse qir-rt library as much as possible while extending it with the callbacks specified above.
+
+__Conditionals on measurements__: The Resource Tracer will execute LLVM IR's branching structures "as is", depending on
+ the values of the corresponding variables at runtime. To enable estimation of branches that depend on a measurement
+ result, the source Q# program will have to be authored in such a way that the Q# compiler will translate the
+ conditionals into `__quantum__qis__apply_if*` calls. The tracer will add operations from _both branches_ into the

```

Comment:
> The tracer will add operations from both branches into the layers it creates to compute the upper bound estimate.

I don't understand this sentence until I know what the _layer_ is.

Reply:
Resolved now?